File: TicTacToe/State.cs
using System;

namespace TicTacToe
{
    public class State
    {
        public static State
            X = new State(true, false),
            O = new State(false, true),
            Empty = new State(false, false);

        public bool Is(State other)
        {
            return isX == other.isX && isO == other.isO;
        }

        private bool isX, isO;

        private State(bool isX, bool isO)
        {
            if (isX && isO)
            {
                throw new ArgumentException("A state cannot be both X and Y");
            }

            this.isX = isX;
            this.isO = isO;
        }
    }
}

File: proj/state.cc

#include "state.h"

// public

const State
    State::X{true, false},
    State::O{false, true},
    State::Empty{false, false};

State::State()
{
    *this = State(false, false);
}

bool State::operator==(const State &other) const
{
    return isX == other.isX && isO == other.isO;
}

// protected

// private

State::State(bool isX, bool isO) :
    isX{isX},
    isO{isO}
{
    if (isX && isO)
    {
        throw "A state cannot be both X and Y";
    }
}
